lib/faraday/connection.rb

```diff
@@ -384,11 +388,30 @@ def build_request(method)
       Request.create(method) do |req|
         req.params  = self.params.dup
         req.headers = self.headers.dup
-        req.options = self.options.merge(:proxy => self.proxy)
+        req.options = self.options
         yield(req) if block_given?
       end
     end
 
+    def proxy_allowed?(url)
+      return true if @no_proxy.nil?
+      uri = Utils.URI(url)
+      @no_proxy.none? do |no_proxy_domain|
+        no_proxy_uri = Utils.URI(no_proxy_domain)
+        # assume strings without scheme are of http
+        if no_proxy_uri.class == URI::Generic && !no_proxy_domain.empty?
+          no_proxy_uri = Utils.URI("http://#{no_proxy_domain}")
+        end
+        if no_proxy_uri.port == 80
+          # domain matching is fine
+          uri.host.downcase.end_with?(no_proxy_uri.host)
+        else
+          # need to match ports exactly
+          "#{uri.host.downcase}:#{uri.port}".end_with?("#{no_proxy_uri.host}:#{no_proxy_uri.port}")
+        end
```

Comment:
I copied this comparison code from the previous [PRs](https://github.com/lostisland/faraday/pull/637/files#diff-fc586c64171977c4441ff2d0f8bba0e2R115).

For your first comment, if the port is the default (port 80) we don't want to compare port numbers, but only compare domain names.  Without this the [test_proxy_allowed_when_port_is_not_set_in_no_proxy_list](
https://github.com/lostisland/faraday/pull/658/files#diff-6c3548c0c264b7967fcecf8c452ee633R410) fails.

For your second comment, we need the end_with? because the `*.` was `gsub`ed out of the no_proxy string assuming that we want to disallow any domain ending.  For example, a certain shape company might want to exclude internal domains ending with `*.circle` from being proxied.  They could just add `circle` to their no_proxy environment variable then.  This is how it works with other tools such as `curl`.  I can't find a spec or RFC on no_proxy environment variable usage, but this seems to be the general practice.